Add support for Clang-Tidy and additional Clang warnings

agarny opened the pull request:
Addresses issue #283.

Clang-Tidy and the additional Clang warnings are used for our `src` and `tests` directories. However, when it comes to the `tests` directory, we had to consider `gtest/` as a system header prefix, to prevent Clang from generating an awful lot of warnings. In the same way, we had to disable some warnings in Clang-Tidy, so that we can "safely" use  `TEST()` and `EXPECT_EQ()`. Also, note that in order to use Clang-Tidy, we had to bump our minimum requirement for CMake. So, now, we need at least CMake 3.6 (while 3.5.2 before).

Note that Clang-Tidy and the additional Clang warnings only make sense on our macOS build machine. Actually, Clang-Tidy could be installed on our Windows and Linux build machines too, if we wanted, but Clang is not used on those machines anyway, so probably best to just stick to our macOS build machine to get a full report of what could be improved in our codebase, if anything, not least since our macOS build machine is the fastest of all three.

The bottom line is that we are now checking an awful lot of things, which should make our codebase pretty solid.

Otherwise, I have taken that opportunity to also:
 - clean up various bits of codes here and there;
 - move the search for LibXml2 from `CMakeLists.txt` to `cmake/common.cmake`, which I think is where it really ought to be?
 - upgrade our version of GTest. We are now using version 1.8.1. This wasn't strictly required, but at some point I thought upgrading GTest might prevent the thousands of warnings I was getting, but it didn't make any difference. Yet, I had upgraded, so we might as well use the latest version of GTest.

agarny committed:
Look for Clang-Tidy and, if available, use it to perform static analysis on our codebase.

Note: this requires CMake 3.6, hence bumping up our CMake version
requirement.

agarny committed:
Clang-Tidy: don't use cppcoreguidelines-slicing.

Indeed, we are “OK” with padding issues not being reported (see
discussion at https://github.com/cellml/libcellml/pull/282).

agarny committed:
Make CMake 3.14 happy.

agarny committed:
Clang-Tidy: treat warnings as errors, if required.

agarny committed:
Consistency clean up.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Optimise the passing of arguments.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

FWIW, Clang-Tidy is not happy with BAD_CAST (and right so), but not
happy with replacing it with reinterpret_cast(). So, still do it but
ask Clang-Tidy not to warn us about reinterpret_cast().

agarny committed:
Clang-Tidy: allow slicing.

And thus prevent some warnings about some bytes being discarded in some
operator=() calls.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
No need for those `assert()` calls.

Indeed, we are using `enum class`, so we ought to be fine.

agarny committed:
Added `llvm-*` to our list of Clang-Tidy warnings.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Clang-Tidy: added Google-specific warnings (except TODO readability one).

The TODO readability one requires specifying a user, e.g. TODO(John)
while we are probably happy with just TODO.

agarny committed:
Revert "Clang-Tidy: added Google-specific warnings (except TODO readability one)."

This reverts commit 746d8a7acb2bc989d3e29055234592f183e957e1.

agarny committed:
Clang-Tidy: added Google-specific warnings (except TODO readability one).

The TODO readability one requires specifying a user, e.g. TODO(John)
while we are probably happy with just TODO.

agarny committed:
Clang-Tidy: want all readability warnings.

braces-around-statement, implicit-bool-conversion and named-parameter
didn’t use to be selected.

agarny committed:
Some minor cleaning up.

agarny committed:
Make Clang-Tidy happy.

agarny committed:
Clang-Tidy: removed google-runtime-references.

Indeed, we want to be able to pass some “output” parameters by
reference.

agarny committed:
Clang-Tidy: added some Fuchsia-specific warnings.

Didn't add fuchsia-default-arguments and
fuchsia-statically-constructed-objects since we are using them quite a
bit.

agarny committed:
Clang-Tidy: added hicpp-specific warnings.

agarny committed:
Clang compiler: add a few more warnings.

Use additional warnings in case we are building libCellML using Clang.
Those warnings were taken from Qt Creator’s Clang Code Model when using
the “Clang-only checks for almost everything” diagnostic
configurations. However, we “added” the following ones:
-Wno-documentation-unknown-command, -Wno-padded and
-Wno-reserved-id-macro. -Wno-documentation-unknown-command is needed to
allow to document things the way we want. -Wno-padded is needed so that
we don’t worry about padding (and is consistent with what we have in
Clang-Tidy). -Wno-reserved-id-macro is needed to prevent some warnings
in system header files.

agarny committed:
Make Clang happy.

agarny committed:
Some minor cleaning up.

agarny committed:
Printer class: removed some methods from the private section of the Printer class.

agarny committed:
Use the same number of spaces as in our other CMake files.

agarny committed:
Consistency cleaning up.

agarny committed:
Moved our search for LibXml2 to cmake/environmentchecks.cmake.

agarny committed:
Made a CMake function of our Clang and Clang-Tidy configuration.

agarny committed:
CMake: call configure_clang_and_clang_tidy_settings() for our various tests.

agarny committed:
Some minor cleaning up.

agarny committed:
Upgraded GTest to version 1.8.1.

I was hoping that the latest version GTest wouldn’t generate so many
warnings, but it still does. Yet, it doesn’t harm upgrading, so we may
as well do it.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Added support for additional Clang warnings and Clang-Tidy to our tests.

Clang:
 - Consider any included file under “gtest/“ as a system header file
and therefore ignore compiler issues with it.

Clang-Tidy:
 - cppcoreguidelines-special-member-functions and
hicpp-special-member-functions: don’t use it for our tests, so that we
can call `TEST()` without any problems.
 - cppcoreguidelines-pro-type-vararg and hicpp-vararg: don’t use them
for our tests, so that we can call `EXPECT_EQ()` without any problems.

agarny committed:
Make Clang and Clang-Tidy happy.

agarny committed:
Some minor cleaning up to make the code consistent with the rest of the codebase.

agarny committed:
Test: this should make our macOS build fail (due to Clang-Tidy / our additional Clang warnings).

agarny committed:
Make Clang and Clang-Tidy happy.

agarny commented:
Argh, our Linux build machine only has CMake 3.5.2 installed on it. @hsorby, I thought it was bioeng49, but it clearly isn't since bioeng49 has CMake 3.12.4. So, which machine is used for building libCellML on Linux? In fact, could you upgrade its copy of CMake?

Regarding our macOS build machine, I also thought that it was bn337051, but once again it doesn't seem to be the case since it seems to be running macOS 10.12...? I installed Clang-Tidy on bn337051, but it looks like it might have been unnecessary. Still, for this PR to make any sense, we need Clang-Tidy installed on the macOS build machine used to build libCellML. Again, @hsorby, it would be nice to have that done.

In summary:
 - [ ] Have our Linux build machine use at least CMake 3.6.
 - [ ] Install Clang-Tidy on our macOS build machine (using `brew install llvm` and then symlinking `clang-tidy` (and `clang-format`) to `/usr/local/bin`).
 - [ ] Repeat commit 89cbd45 to make sure that our macOS build does indeed fail.
 - [ ] Make Clang and Clang-Tidy happy by committing something like 029a7e4.

agarny committed:
Make Clang and Clang-Tidy happy.

Dealt with readability-isolate-declaration issues.

agarny committed:
Make it clearer what we use / don't use in Clang and Clang-Tidy.

agarny committed:
Swapped the declaration of version() and versionString().

agarny committed:
Made Clang and Clang-Tidy happy.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny commented:
Would be nice to have something done about this PR... especially with Keri now working on libCellML. I would also like to have `clang-format` used to ensure that our code is consistent throughout (see issue #284).

nickerso commented:
Just to be clear, we're waiting on some upgrades to the build machines to get things playing nice with these changes, right?

hsorby commented:
I believe you are right.

agarny commented:
That is correct. The version of CMake on the Linux slave is too old for this PR to pass, and on the macOS slave we should have `clang-tidy` installed so that this PR can actually be useful.

I thought that those slaves were `bioeng49` and `bn337051`, but they are not. `bioeng49` runs Ubuntu 16.04 LTS while the Linux slave used by libCellML runs Ubuntu 14.04 LTS (?!). As for the macOS slave used by libCellML, it seems to be running macOS Sierra (10.12) while `bn337051` runs macOS High Sierra (10.13).

So, in the end, I have no idea what/where those slaves are... and therefore can't update them...

agarny committed:
Some minor cleaning up.

- Renamed configure_clang_and_clang_tidy_settings() to
configure_clang_and_clang_tidy().
- Added a leading “_” to our configure_clang_and_clang_tidy() parameter
and its local variables. (This seems to be the convention, although not
the one I would have chosen.)

agarny committed:
Some minor cleaning up.

agarny committed:
Don't allow shadow variables.

agarny committed:
Some minor cleaning up.

agarny committed:
Don't ignore enum issues.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

hsorby commented:
The build slaves have now been updated and all pertinent builds have been triggered.  Failures now indicate true failures that need to be fixed.  This (I'm fairly confident) is the only PR needing changes.

agarny commented:
Hmm, not quite sure how I missed that one, but I can indeed confirm the same build error on my local Ubuntu 18.04 LTS box. Ok, consider it fixed... soon...

agarny committed:
Version test: fixed a small issue with our Linux build.

nickerso commented:
This one is good to merge in now, right? is it going to cause major headaches in the other open pull requests (especially #304) ?

agarny commented:
> This one is good to merge in now, right? is it going to cause major headaches in the other open pull requests (especially #304) ?

As far as I am concerned, it is indeed ready to merge. And, yes, it's likely to cause some headache with all pending PRs...

hsorby commented:
First in first served?  That is first to get their PR approved and merged passes on the burden of conflict resolution to PRs not yet approved.

agarny committed:
Merge branch 'develop' into issue283.

hsorby commented on cmake/environmentchecks.cmake:
```diff
@@ -14,21 +14,54 @@
+  if (LibXml2_FOUND)
+    set(HAVE_LIBXML2_CONFIG TRUE)
+  else ()
+    find_package(LibXml2 REQUIRED)
```
Can we add in here:

`unset(LibXml2_DIR CACHE)`

agarny replied:
There are several calls to `find_package(LibXml2...)`. So, are you meaning that we should that `unset()` call between `set(HAVE_LIBXML2_CONFIG FALSE)`and `if(MSVC)`?

agarny committed:
Some minor cleaning up.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Test: just for clarity.

agarny committed:
Revert "Test: just for clarity."

This reverts commit ed9b3cbe845e73a35ebeb31ae38c092883899497.

Clang-Tidy (rightly) reports that it’s “redundant string
initialization”…

agarny committed:
Make ClangFormat happy.

agarny committed:
Some minor cleaning up.

agarny commented:
@hsorby, would you know why the Python tests fail on the macOS slave? FWIW, those tests work on my Mac...

hsorby commented:
Yes, the macOS slave is running SWIG 4.0.  So we need to look at that.

hsorby commented:
The new SWIG 4.0 has changed the way it deals with rpath I think so we need to upgrade the handling of the bound dynamic libraries.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Merge branch 'develop' into issue283

hsorby committed:
Merge branch 'develop' into ctidy_i283

hsorby committed:
Minor tidy up for clang-tidy around CMake if statements.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Merge branch 'develop' into issue283

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Added XmlNode::isMathmlElement() and XmlAttribute::isCellmlType().

agarny committed:
Some minor cleaning up,

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: fixed our private gatherMathBvarVariableNames() method.

It would always check only the first (grand)child, not taking into
account the fact that it might not be of the expected type.

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: improved the validation of MathML elements.

hsorby committed:
Restore some lost CMake logic when finding libXml2.

hsorby commented:
I think this one is good to merge now.

agarny commented:
FWIW, I want to go through this one more time (today). There have been quite a few `git merge`s lately. So, I want to make sure that it all looks good.

agarny commented:
Actually, @hsorby, don't we want to create a test for Clang-Tidy in the same that you created one for ClangFormat?

agarny committed:
Merge branch 'develop' into issue283.

agarny committed:
Renamed some instances of "clang-format" to "ClangFormat".

agarny committed:
Some minor cleaning up.

agarny committed:
Partially revert commit d5b6d4e.

Oops, those were actually ClangFormat commands…!

hsorby commented:
The problem with the clang-tidy is that it is a compiler level change.  It is not really external.  The clang tidy then becomes part of the standard warnings and errors that we already deal with.  So I think we just add it in when available the Buildbot testing will do the same and our checks should continue as they currently are with no further changes necessary.

agarny commented:
> The problem with the clang-tidy is that it is a compiler level change. It is not really external. The clang tidy then becomes part of the standard warnings and errors that we already deal with. So I think we just add it in when available the Buildbot testing will do the same and our checks should continue as they currently are with no further changes necessary.

Ok, fair enough.

Otherwise, I have just gone through this PR one more time and I am happy with it. So, up to the reviewers now... :)

agarny commented:
From what I can tell, both @hsorby and @nickerso are OK with this PR. So, either of you can merge it in...
